compiler/rustc_expand/src/expand.rs

```diff
@@ -546,7 +547,12 @@ impl<'a, 'b> MacroExpander<'a, 'b> {
                         })
                         .unwrap_or_default();
 
-                    self.collect_invocations(fragment, &derive_placeholders)
+                    let (fragment, collected_invocations) =
+                        self.collect_invocations(fragment, &derive_placeholders);
+                    // Any derive invocations associated with this macro invocation should be
+                    // expanded *before* any macro invocations collected from the output fragment
```

Comment:
I don't understand this.

The order of entries in the top level `invocations` vector is non-significant, it's a job queue - entries can return to it if they are not ready to be resolved or expanded so the exact expansion order is unspecified in general.
So this PR changes one unpredictable order to another?

Once `#[derive(Trait1, Trait2)] ITEM` produced `ITEM`, `placeholder(Trait1)` and `placeholder(Trait2)` these three items part their ways from each other and can be expanded independently in any order.
The existing code chooses the tentative order `ITEM` -> `placeholder(Trait2)` -> `placeholder(Trait1)`, but it may change later in invocations in `ITEM` are not yet ready.
The new code chooses the tentative order `placeholder(Trait1)` -> `placeholder(Trait2)` -> `ITEM` (which may also change, but only inside `ITEM` because the two placeholders here are always ready by construction).

Am I missing anything?

Reply:
My reasoning was that we should be treating `#[derive]` like a normal macro as much as possible. While the special delayed-resolution behavior of multiple `#[derive]` attributes (compared to normal attributes) means that we can't guarantee this, I think it makes sense to expand derives left-to-right when possible.

By doing so, we'll make the order of messages printed by derive macros more comprehensible in some cases. Even though the order is unspecified, I don't think we should intentionally pick an order (when everything can be resolved immediately) that's inconsistent with how users normally expect attribute macros to work.

Reply:
You can see the "FIXME: Consider using the derive resolutions ..." near the changed code.
I've just realized that if it's fully implemented, then we'll stop enqueuing derives entirely, just expand them immediately in place (we have all the data necessary to do that).
In that case it'll indeed match the new order implemented in this PR.

I also agree the new order is more intuitive even if it's not specified, so it seems fine to change it.

Reply:
Could you change the "should be expanded *before*" comment to something less prescriptive? (The new order is a choice, not necessity.)

Reply:
https://github.com/rust-lang/rust/pull/82907 was merged so addressing the FIXME is actually pretty close to the top of my work queue.